### PR TITLE
Ports: Cave Story (Aeon Genesis Translation)

### PR DESCRIPTION
**Pull request by gloof11**

### CODE REVIEW

**Comment by AtkinsSJ** [comment]
OK, this still needs a little more git-fiddling. Those three commits really want to just be one. There are a couple of ways to do that but the simplest is probably:
```bash
# First make sure the branch is selected. You can probably skip it.
> git checkout cavestory
# Now we rebase with `-i` to let us modify history.
> git rebase -i master
# Now, you should see the three commits in an editor. Change the words in front of the bottom two from "pick" to "fixup". Then save and exit. (fixup means "apply these changes to the previous commit"). If you make any mistakes editing this file, just select everything and delete it, and git will 
# Finally, push the changes.
> git push --force-with-lease
```

There's also this video that we regularly share with people: [Rewriting Git History](https://www.youtube.com/watch?v=ElRzTuYln0M)

It gets easier, don't worry. :^) I made many messes at first.

File: Ports/cavestory/package.sh [hunk 1]
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env -S bash ../.port_include.sh
+
+port="cavestory"
```
Comment (gmta) [inline Ports/cavestory/package.sh:3]: All strings without interpolation should use single quotes. For example, in this case:

```
port='cavestory`
```

File: Ports/cavestory/package.sh [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+auth_type='sha256'
+depends=(
+  "gcc"
+  "cmake"
+  "SDL2"
+  "SDL2_ttf"
+  "libpng"
+  "SDL2_mixer"
+  "libjpeg"
+  "SDL2_image"
```
Comment (gmta) [inline Ports/cavestory/package.sh:15]: 1. Please use 4 spaces as indentation
2. All these entries should use single quotes
3. Please put them in alphabetical order

File: Ports/cavestory/package.sh [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+build () {
+  cd $workdir/build; cmake -DCMAKE_BUILD_TYPE=Release .. 
+  make
+}
+
+pre_install () {
+  curl https://www.cavestory.org/downloads/cavestoryen.zip -O 
+  unzip cavestoryen.zip
```
Comment (gmta) [inline Ports/cavestory/package.sh:34]: Download + unzip is exactly what the `files` array is for - can't you use that?

File: Ports/cavestory/package.sh [hunk 4]
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env -S bash ../.port_include.sh
+
+port="cavestory"
+version="2.6.5-1"
+files="https://github.com/nxengine/nxengine-evo/archive/refs/heads/master.zip master.zip d013acbf399938b5d3d10fd4d37817b630559e2b24b90576cde6f88d445ea150"
+auth_type='sha256'
+depends=(
+  "gcc"
+  "cmake"
```
Comment (gmta) [inline Ports/cavestory/package.sh:9]: Does Cave Story really depend on gcc and cmake? Or are these host tools required for the build? Because if it's the latter, then these should not be in here.

File: Ports/cavestory/patches/0003-Updated-the-.gitignore-to-ignore-unpacked-game-extra.patch [hunk 5]
```diff
@@ -0,0 +1,54 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: gloof11 <gloof11@proton.me>
+Date: Mon, 26 Jun 2023 19:37:55 +0900
+Subject: [PATCH] Updated the .gitignore to ignore unpacked game extract files. Removed hardware acceleration from SDL_CreateRenderer
```
Comment (gmta) [inline Ports/cavestory/patches/0003-Updated-the-.gitignore-to-ignore-unpacked-game-extra.patch:4]: This should be two separate patches

**Comment by gmta** [comment]
I get the following error:
```
[cavestory/build] CMake Error at /path/to/serenity/Toolchain/Local/cmake/share/cmake-3.26/Modules/FindPackageHandleStandardArgs.cmake:230 (message):
[cavestory/build]   Could NOT find SDL2 (missing: SDL2_LIBRARY)
[cavestory/build] Call Stack (most recent call first):
[cavestory/build]   /path/to/Projects/serenity/Toolchain/Local/cmake/share/cmake-3.26/Modules/FindPackageHandleStandardArgs.cmake:600 (_FPHSA_FAILURE_MESSAGE)
[cavestory/build]   cmake/Modules/FindSDL2.cmake:180 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
[cavestory/build]   CMakeLists.txt:55 (find_package)
```

File: Ports/cavestory/package.sh [hunk 6]
```diff
@@ -0,0 +1,49 @@
+#!/usr/bin/env -S bash ../.port_include.sh
+
+port="cavestory"
+version="2.6.5-1"
+files="https://github.com/nxengine/nxengine-evo/archive/refs/heads/master.zip master.zip d013acbf399938b5d3d10fd4d37817b630559e2b24b90576cde6f88d445ea150"
```
Comment (timschumi) [inline Ports/cavestory/package.sh:5]: Please pin this to a specific revision or tag instead of `master`. Otherwise, the hash will no longer match as soon as the repository is updated.

File: Ports/cavestory/patches/ReadMe.md [hunk 7]
```diff
@@ -0,0 +1,9 @@
+# Patches for cavestory on SerenityOS
+
+## `0001-Removed-references-to-fstat64.patch`
+
+Removed references to fstat64 Fixed how timeIf __FreeBSD__, __APPLE__, __x86_64__, __ppc64__, __CYGWIN__, or __HAIKU__ is not defined it will use "fstat64" to get file status. Currently SerenityOS does not support fstat64, so the code was modified to not perform the check, and instead use "fstat" is generated
```
Comment (AtkinsSJ) [inline Ports/cavestory/patches/ReadMe.md:5]: This looks a bit muddled, did you combine a couple of sentences while fixing the git mess?

File: Ports/cavestory/package.sh [hunk 8]
```diff
@@ -0,0 +1,47 @@
+#!/usr/bin/env -S bash ../.port_include.sh
+
+port='cavestory'
+version='2.6.5-1'
+files=(
+    'https://github.com/nxengine/nxengine-evo/archive/refs/tags/v2.6.5-1.zip v2.6.5-1.zip a3a1a802eda46f9922065a35624ba94e3823649a7e71012e479ed91fcdb50255'
```
Comment (timschumi) [inline Ports/cavestory/package.sh:6]:
```suggestion
    "https://github.com/nxengine/nxengine-evo/archive/refs/tags/v${version}.zip v${version}.zip a3a1a802eda46f9922065a35624ba94e3823649a7e71012e479ed91fcdb50255"
```

File: Ports/cavestory/patches/0001-Removed-references-to-fstat64.patch [hunk 9]
```diff
@@ -0,0 +1,56 @@
...
+-#endif
+ #endif
+     throw spdlog_ex("Failed getting file size from fd", errno);
+ }
+@@ -271,8 +263,6 @@ inline int utc_minutes_offset(const std::tm &tm = details::os::localtime())
+     }
+     return offset;
+ #else
+-
+-#if defined(sun) || defined(__sun) || defined(_AIX) || defined(__VITA__) || defined(__SWITCH__)
```
Comment (timschumi) [inline Ports/cavestory/patches/0001-Removed-references-to-fstat64.patch:42]: Same here, but with `|| defined(__serenity__)`.

**Review by timschumi — CHANGES_REQUESTED** [review]

File: Ports/cavestory/patches/0001-Removed-references-to-fstat64.patch [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+diff --git a/deps/spdlog/details/os.h b/deps/spdlog/details/os.h
+index 57df5a85a0a5b601b17121f1f1f17c536496b132..67769ce1f3f2608dbaa79cd8cf76f21559d39746 100644
+--- a/deps/spdlog/details/os.h
++++ b/deps/spdlog/details/os.h
+@@ -226,21 +226,13 @@ inline size_t filesize(FILE *f)
+ 
+ #else // unix
+     int fd = fileno(f);
+-// 64 bits(but not in osx or cygwin, where fstat64 is deprecated)
+-#if !defined(__FreeBSD__) && !defined(__APPLE__) && (defined(__x86_64__) || defined(__ppc64__)) && !defined(__CYGWIN__) && !defined(__HAIKU__)
```
[the remark this answers is not in the document]
Comment (gloof11) [reply Ports/cavestory/patches/0001-Removed-references-to-fstat64.patch:19]: I didn't discover that __serenity__ is recognized as a valid define until recent. Will update in the next commit.

File: Ports/cavestory/package.sh [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    'SDL2_ttf'
+)
+workdir="nxengine-evo-${version}"
+launcher_name='Cave Story'
+launcher_category='Games'
+launcher_command='/usr/local/bin/nxengine-evo'
+icon_file='platform/switch/icon.jpg'
+
+post_fetch () {
+  mkdir -p ${workdir}/build
```
Comment (ADKaster) [inline Ports/cavestory/package.sh:24]: four spaces indentation should be consistent everywhere.

File: Ports/cavestory/package.sh [hunk 12]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+build () {
+  cd ${workdir}/build
+  cmake -DCMAKE_BUILD_TYPE=Release ..
```
Comment (ADKaster) [inline Ports/cavestory/package.sh:29]: You don't need the post-fetch step or any of the `cd workdir` steps.

`run cmake -B build "${configopts[@]}"` as `configure()` should do the trick if you set `useconfigure` to true.

and `run make -C build` as `build()` should also do the trick. see other cmake ports for an example.

How does this get access to our CMakeToolchain file? As written it looks like this will build for the host.

the release build type and CMakeToolchain file should be included in `configopts`

File: Ports/cavestory/package.sh [hunk 13]
```diff
@@ -0,0 +1,49 @@
...
+build () {
+  cd ${workdir}/build
+  cmake -DCMAKE_BUILD_TYPE=Release .. 
+  make
+}
+
+pre_install () {
+  mv CaveStory/Doukutsu.exe ${workdir}/
+  mv CaveStory/data/* ${workdir}/data/
+  rm -rf CaveStory
```
Comment (ADKaster) [inline Ports/cavestory/package.sh:36]: Why are we deleting this data?

File: Ports/cavestory/package.sh [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+  export PATH=$PORT_BUILD_DIR/${workdir}:$PATH
+  cd ${workdir}
+  nxextract
```
Comment (ADKaster) [inline Ports/cavestory/package.sh:39]: `run nxextract`

File: Ports/cavestory/package.sh [hunk 15]
```diff
@@ -0,0 +1,49 @@
...
+  mv CaveStory/data/* ${workdir}/data/
+  rm -rf CaveStory
+  export PATH=$PORT_BUILD_DIR/${workdir}:$PATH
+  cd ${workdir}
+  nxextract  
+}
+
+install () {
+  cd ${workdir}/build 
+  make prefix=/usr/local DESTDIR=$SERENITY_INSTALL_ROOT install
```
Comment (ADKaster) [inline Ports/cavestory/package.sh:44]: `run make -C build install`

Those other variables should already be set by `.port_include.sh`

Perhaps prefix is required. but DESTDIR is already set

File: Ports/cavestory/package.sh [hunk 16]
```diff
@@ -0,0 +1,49 @@
...
+}
+
+uninstall () {
+  cd ${workdir}/build
```
Comment (ADKaster) [inline Ports/cavestory/package.sh:48]: This uninstall doesn't do anything.

**Review by ADKaster — CHANGES_REQUESTED** [review]
This port looks pretty straightforward now. Please resolve everyone's comments once you've addressed them.

File: Ports/cavestory/package.sh (hunk 13 above)
Comment (gloof11) [reply Ports/cavestory/package.sh:36]: The "rm -rf CaveStory" is there to delete the unzipped folder, now that all of the game's data was extracted from it. Does Serenity automatically clean this folder up? If not there would just be a folder sitting there taking up space.

Comment (ADKaster) [reply Ports/cavestory/package.sh:36]: That's only one folder though. We download tarballs and extract all sorts of files as part of the Toolchain build, project build, and every port build. None of it is cleaned up unless you take action yourself. And it's all downloaded into the Build/<arch> folder anyway 

What if I run the build and install steps again? It will fail unless I do a full clean rebuild of the port.

Comment (gloof11) [reply Ports/cavestory/package.sh:36]: You're right, but I feel it's best to not have unnecessary files, even if other ports do it.

File: Ports/cavestory/package.sh (hunk 15 above)
Comment (gloof11) [reply Ports/cavestory/package.sh:44]: Trying to run it without DESTDIR will attempt to install it to the host's /usr/local/bin.

File: Ports/cavestory/package.sh (hunk 13 above)
Comment (gmta) [reply Ports/cavestory/package.sh:36]: For reproducibility and debugging purposes, it's best to leave the files after running the steps. This is the same for all ports, so we should not diverge here.

If anyone's concerned with disk space usage, an option could be trivially added to the invocation of `package.sh` to clean up afterwards (but let's not do that in this PR).

File: Ports/cavestory/package.sh (hunk 15 above)
Comment (gmta) [reply Ports/cavestory/package.sh:44]: I think you need to pass `"${makeopts[@]}"` as an argument to `make` to make that work. That's the best way to do it since it will always use the configured make options in the future.

File: Ports/cavestory/package.sh (hunk 16 above)
Comment (gmta) [reply Ports/cavestory/package.sh:48]: This is not resolved - `uninstall` should be removed.

File: Ports/cavestory/package.sh (hunk 15 above)
Comment (ADKaster) [reply Ports/cavestory/package.sh:44]: ${installopts[@]} in this case iirc

Comment (gloof11) [reply Ports/cavestory/package.sh:44]: "${makeopts[@]}" by itself will not work. However, setting makeopts to "DESTDIR=$DESTDIR" fixes it.

File: Ports/cavestory/package.sh (hunk 12 above)
Comment (gloof11) [reply Ports/cavestory/package.sh:29]: I'm not sure how, but it certainly gets access to it. I don't have CMake installed but looking at the source code in CMakeLists.txt, it does nothing if the CMAKE_TOOLCHAIN_FILE environment variable is set on the host.

File: Ports/cavestory/package.sh (hunk 15 above)
Comment (gloof11) [reply Ports/cavestory/package.sh]: > ${installopts[@]} in this case iirc

${installopts[@]} will also attempt to install to the host's /usr/local/bin

Comment (gmta) [reply Ports/cavestory/package.sh]: @ADKaster @gloof11 You are both right.

https://github.com/SerenityOS/serenity/blob/bfd6deed1e860ce9516c7ca1613150a2c2c75f75/Ports/.port_include.sh#L477

Maybe we should have included `DESTDIR` in `installopts` instead of adding it here as a separate argument? 🤔

**Comment by gmta** [comment]
This is still failing for me on the SDL2 error mentioned above. You should make sure that you don't have SDL2 development headers on your host system before retrying.

File: Ports/cavestory/package.sh [hunk 17]
```diff
@@ -0,0 +1,38 @@
...
+    "-DSDL2_IMAGE_INCLUDE_DIR=${SERENITY_INSTALL_ROOT}/usr/local/include/SDL2"
+    "-DSDL2_MIXER_INCLUDE_DIR=${SERENITY_INSTALL_ROOT}/usr/local/include/SDL2"
+)
+
+configure() {
+    run cmake -B build "${configopts[@]}"
+}
+
+build () {
+    run make -C build
```
Comment (gmta) [inline Ports/cavestory/package.sh:33]:
```suggestion
    run make -C build "${makeopts[@]}"
```

This will make sure we build in parallel, since `makeopts` defaults to have the `-j` argument in there.

**Review by gmta — APPROVED** [review]
